backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/query/impl/LuceneSearcherImpl.java

```diff
@@ -147,6 +148,37 @@ private LuceneCollectors buildCollectors(IndexSearcher indexSearcher, int offset
 		return luceneCollectorsBuilder.build();
 	}
 
+	private void executeQuery(IndexSearcher indexSearcher, int offset, Integer limit, LuceneCollectors luceneCollectors) throws IOException {
+		boolean timeoutAtZero = checkTimeoutAtZero();
+		if ( timeoutAtZero ) {
+			// in case of timeout at zero, skipping the query
+			return;
+		}
+
+		try {
+			luceneCollectors.collect( indexSearcher, requestContext.getLuceneQuery(), offset, limit );
+		}
+		catch (TimeLimitingCollector.TimeExceededException e) {
+			timeoutManager.forceTimedOut();
+		}
+	}
+
+	private boolean checkTimeoutAtZero() {
```

Comment:
I don't think this was previously implemented in `TimeOutManager.isTimedOut()`. See `QueryHits#isImmediateTimeout` and its usage `QueryHits:236`.

Reply:
Well, there's something strange, because isTimedOut() will definitely throw an exception the first time it detects a timeout:

```
	@Override
	public boolean isTimedOut() {
		if ( timeout == null ) {
			return false;
		}
		if ( timedOut ) {
			return true;
		}
		return isTimedOut( System.nanoTime() );
	}

	private boolean isTimedOut(long currentTime) {
		if ( timeout == null ) {
			return false;
		}
		if ( timedOut ) {
			return true;
		}
		else {
			final long elapsedTime = currentTime - start;
			timedOut = elapsedTime > timeout;
			if ( this.type != Type.LIMIT && timedOut ) {
				throw timeoutExceptionFactory.createTimeoutException(
						"Full-text query took longer than expected (in microsecond): " + TimeUnit.NANOSECONDS.toMicros( elapsedTime ),
						String.valueOf( query )
				);
			}
			return timedOut;
		}
	}
```

In any case... moving timeout management to the timeout manager would be a good idea, I think.

Reply:
Yep. I does make sense to call the same `isTimedOut() ` we periodically call during the loading to check if a timeout is already present at the very beginning as well (time out at zero case).